core/server/controllers/frontend/index.js

```diff
@@ -66,11 +66,21 @@ frontendControllers = {
                 return next();
             }
 
-            // If we're ready to render the page but the last param is 'edit' then we'll send you to the edit page.
+            // CASE: we only support /:slug format for pages
+            if (post.page && post.url !== req.path) {
+                return next();
+            }
+
+            // CASE: last param is of url is /edit, redirect to admin
             if (lookup.isEditURL) {
                 return res.redirect(config.paths.subdir + '/ghost/editor/' + post.id + '/');
             }
 
+            // CASE: permalink is not valid anymore, we redirect him to the correct one
+            if (post.url !== req.path) {
+                return res.redirect(post.url);
```

Comment:
This should, in my opinion, use a 301 redirect, rather than express's default of a 302.